### ISSUE TITLE
A firmware update will not install on my Lenovo Thinkpad X1 Carbon 7th running Ubuntu 21.04

### ISSUE THREAD

**Issue by cdysthe** [fwupd/firmware-lenovo#123]
**Describe the bug**
The 'Ubuntu Software' application tells me me I have a firmware update available (screenshot below). I have installed many of those on this laptop without problems but this time when I click "Install" the update notification disappears and then shows up again as if I haven't done anything. I can keep clicking 'Install' but it keeps coming back after a couple of seconds. I then thought I would reboot and maybe the update would install before boot as it has in the past. Nothing, just a normal boot, but I notice a lockup/pause for a couple of seconds during shutdown I haven't noticed before.

I also tried using fwupd but It tells me I have an update, I choose to install it and it's being downloaded. Now I am getting a reboot y/n option so I reboot. Same thing, nothing other than the pause during shutdown. I boot normally again and still have the same update available right after I log in to my user account.

**Steps to Reproduce**
Either try to install the available update in Ubuntu Software or do a  'sudo fwupdmgr update'

**Expected behavior**
The available update to be installed and applied after a reboot either from 'Ubuntu Software' ot fwupd. fwupd asks to be allowed to reboot.

**fwupd version information**
```shell
$ fwupdmgr --version
client version:	1.5.8
compile-time dependency versions
gusb:	0.3.5
daemon version:	1.5.8
``` 
Please note how you installed it (`apt`, `dnf`, `pacman`, source, etc):

**fwupd device information**
Please provide the output of the fwupd devices recognized in your system.

```shell
$ fwupdmgr get-devices --show-all-devices
20R1S04000
│
├─Thunderbolt host controller:
│     Device ID:          0109c04bcfd8f359899bdd2357e9f1e16db5ab5a
│     Summary:            Unmatched performance for high-speed I/O
│     Current version:    44.00
│     Vendor:             Lenovo (THUNDERBOLT:0x0109|TBT:0x0109)
│     GUIDs:              74cdddba-f7b6-5a31-a3dc-64186b5a5491 ← THUNDERBOLT\VEN_0109&DEV_1801&REV_00
│                         18bb00e3-5d7a-55d3-831e-0e000c184b0e ← THUNDERBOLT\VEN_0109&DEV_1801
│                         c7920601-0cda-507f-851e-92229eb1d370 ← TBT-01091801-native
│                         96c6512b-0138-51da-b02f-f44f2c4c7744 ← TBT-01091801-native-controller0-0
│     Device Flags:       • Internal device
│                         • Updatable
│                         • System requires external power source
│                         • Supported on remote server
│                         • Device stages updates
│   
├─Embedded Controller:
│     Device ID:          d4db94d2324ae6986698e7489b234c0377fd8465
│     Current version:    0.1.15
│     Minimum Version:    0.1.15
│     Vendor:             DMI:LENOVO
│     Update Error:       Cannot update as UEFI Device Firmware [317fd9ef3f069f67047ab1163f8c040c3853cc38] needs reboot
│     GUIDs:              ef793001-2e3e-42fa-ad94-c8d75c62909e
│                         3e51de31-6868-5a27-b91a-f6d89252742f ← UEFI\RES_{EF793001-2E3E-42FA-AD94-C8D75C62909E}
│     Device Flags:       • Internal device
│                         • System requires external power source
│                         • Supported on remote server
│                         • Needs a reboot after installation
│                         • Device is usable for the duration of the update
│                         • Updatable
│   
├─Intel(R) Core™ i5-10210U CPU @ 1.60GHz:
│     Device ID:          4bde70ba4e39b28f9eab1628f9dd6e6244c03027
│     Current version:    0x000000de
│     Vendor:             Intel
│     GUIDs:              b9a2dd81-159e-5537-a7db-e7101d164d3f ← cpu
│                         30249f37-d140-5d3e-9319-186b1bd5cac3 ← CPUID\PRO_0&FAM_06
│                         561403e8-143a-5071-ab09-bf5e4c146983 ← CPUID\PRO_0&FAM_06&MOD_8E
│                         d5092916-1868-5182-8c70-12f83dab6372 ← CPUID\PRO_0&FAM_06&MOD_8E&STP_C
│     Device Flags:       • Internal device
│   
├─Prometheus:
│ │   Device ID:          d432baa2162a32c1554ef24bd8281953b9d07c11
│ │   Summary:            Fingerprint reader
│ │   Current version:    10.01.3158804
│ │   Vendor:             Synaptics (USB:0x06CB)
│ │   Install Duration:   2 seconds
│ │   Serial Number:      137802065387430
│ │   GUIDs:              09e8bf16-3e69-50f5-bb66-c7a040248352 ← USB\VID_06CB&PID_00BD&REV_0000
│ │                       8088f861-6318-5b1e-9ce4-fbddbedb09ac ← USB\VID_06CB&PID_00BD
│ │   Device Flags:       • Updatable
│ │                       • Supported on remote server
│ │                       • Cryptographic hash verification is available
│ │ 
│ └─Prometheus IOTA Config:
│       Device ID:        8dcffb5d059857368caa2b7b22371ebd831b0c0c
│       Current version:  0021
│       Minimum Version:  0021
│       Vendor:           Synaptics (USB:0x06CB)
│       GUIDs:            7c5a1e62-38fa-5859-9337-09dbac6377e4 ← USB\VID_06CB&PID_00BD-cfg
│                         21a62213-7012-552c-9f1a-a51d07c87242 ← USB\VID_06CB&PID_00BD&CFG1_3462&CFG2_0
│       Device Flags:     • Updatable
│                         • Supported on remote server
│     
├─Samsung SSD 970 EVO Plus 1TB:
│     Device ID:          3743975ad7f64f8d6575a9ae49fb3a8856fe186f
│     Summary:            NVM Express Solid State Drive
│     Current version:    2B2QEXM7
│     Vendor:             Samsung Electronics Co Ltd (NVME:0x144D)
│     Serial Number:      S59ANJ0N123561X
│     GUIDs:              0b4d773a-7ac3-58c1-a541-e22ef1cdfe02 ← NVME\VEN_144D&DEV_A808&SUBSYS_144DA801&REV_00
│                         c9d531ea-ee7d-5562-8def-c64d0d144813 ← NVME\VEN_144D&DEV_A808&SUBSYS_144DA801
│                         6e54c992-d302-59ab-b454-2d26ddd63e6d ← NVME\VEN_144D&DEV_A808&REV_00
│                         47335265-a509-51f7-841e-1c94911af66b ← NVME\VEN_144D&DEV_A808
│                         f05ecac9-5cc2-5f77-b23b-a08ad8879ada ← Samsung SSD 970 EVO Plus 1TB
│     Device Flags:       • Internal device
│                         • Updatable
│                         • System requires external power source
│                         • Needs a reboot after installation
│                         • Device is usable for the duration of the update
│   
├─System Firmware:
│ │   Device ID:          09647b198775f1868cc695ec2685477e0964c323
│ │   Current version:    0.1.32
│ │   Vendor:             LENOVO (DMI:LENOVO)
│ │   Update Error:       Cannot update as UEFI Device Firmware [317fd9ef3f069f67047ab1163f8c040c3853cc38] needs reboot
│ │   GUIDs:              765f81e8-cc44-4d09-8fc4-9a47b167166e
│ │                       230c8b18-8d9b-53ec-838b-6cfc0383493a ← main-system-firmware
│ │                       8789d0d6-4e4d-539f-acfc-97447b5ea3d7 ← UEFI\RES_{765F81E8-CC44-4D09-8FC4-9A47B167166E}
│ │   Device Flags:       • Internal device
│ │                       • System requires external power source
│ │                       • Supported on remote server
│ │                       • Needs a reboot after installation
│ │                       • Cryptographic hash verification is available
│ │                       • Device is usable for the duration of the update
│ │                       • Updatable
│ │ 
│ └─UEFI dbx:
│       Device ID:        362301da643102b9f38477387e2193e57abaa590
│       Summary:          UEFI Revocation Database
│       Current version:  238
│       Minimum Version:  238
│       Vendor:           UEFI:Linux Foundation
│       Install Duration: 1 second
│       GUIDs:            14503b3d-73ce-5d06-8137-77c68972a341 ← UEFI\CRT_A9087D1044AD18F7A94916D284CBC01827CF23CD8F60B79072C9CAA1FEF4D649
│                         5971a208-da00-5fce-b5f5-1234342f9cf7 ← UEFI\CRT_A9087D1044AD18F7A94916D284CBC01827CF23CD8F60B79072C9CAA1FEF4D649&ARCH_X64
│                         c6682ade-b5ec-57c4-b687-676351208742 ← UEFI\CRT_A1117F516A32CEFCBA3F2D1ACE10A87972FD6BBE8FE0D0B996E09E65D802A503
│                         f8ba2887-9411-5c36-9cee-88995bb39731 ← UEFI\CRT_A1117F516A32CEFCBA3F2D1ACE10A87972FD6BBE8FE0D0B996E09E65D802A503&ARCH_X64
│       Device Flags:     • Internal device
│                         • Updatable
│                         • Needs a reboot after installation
│     
├─TPM:
│ │   Device ID:          c6a80ac3a22083423992a3cb15018989f37834d6
│ │   Current version:    73.8.17568.5511
│ │   Vendor:             ST Microelectronics (TPM:STM)
│ │   GUIDs:              ff71992e-52f7-5eea-94ef-883e56e034c6 ← system-tpm
│ │                       84df3581-f896-54d2-bd1a-372602f04c32 ← TPM\VEN_STM&DEV_0001
│ │                       bfaed10a-bbc1-525b-a329-35da2f63e918 ← TPM\VEN_STM&MOD_
│ │                       70b7b833-7e1a-550a-a291-b94a12d0f319 ← TPM\VEN_STM&DEV_0001&VER_2.0
│ │                       06f005e9-cb62-5d1a-82d9-13c534c53c48 ← TPM\VEN_STM&MOD_&VER_2.0
│ │   Device Flags:       • Internal device
│ │ 
│ └─Event Log:
│       Device ID:        58bd405f31c48e6eca290b425f530a94c91e955c
│       GUID:             a25657fe-b5dc-5be0-8b78-8b9dfec678ff ← system-tpm-eventlog
│       Device Flags:     • Internal device
│     
├─UEFI Device Firmware:
│     Device ID:          fe0a5afcd3063283e8535294cd48ad186d78707a
│     Current version:    3758752948
│     Minimum Version:    1
│     Vendor:             DMI:LENOVO
│     Update Error:       Cannot update as UEFI Device Firmware [317fd9ef3f069f67047ab1163f8c040c3853cc38] needs reboot
│     GUIDs:              aa6f3567-b1f3-46d1-bf8a-a21f17f286aa
│                         d29f1f35-194b-59d3-88f2-8f67e5fdc1f2 ← UEFI\RES_{AA6F3567-B1F3-46D1-BF8A-A21F17F286AA}
│     Device Flags:       • Internal device
│                         • System requires external power source
│                         • Needs a reboot after installation
│                         • Device is usable for the duration of the update
│                         • Updatable
│   
├─UEFI Device Firmware:
│     Device ID:          60b8b69551637e625ef248a9cff72919c6033f1a
│     Current version:    16785559
│     Minimum Version:    1
│     Vendor:             DMI:LENOVO
│     Update Error:       Cannot update as UEFI Device Firmware [317fd9ef3f069f67047ab1163f8c040c3853cc38] needs reboot
│     GUIDs:              1e856ac4-aa7f-4ea8-8e92-c674c1501cd8
│                         ab573842-6c72-5408-ab02-284d42b21f42 ← UEFI\RES_{1E856AC4-AA7F-4EA8-8E92-C674C1501CD8}
│     Device Flags:       • Internal device
│                         • System requires external power source
│                         • Needs a reboot after installation
│                         • Device is usable for the duration of the update
│                         • Updatable
│   
├─UEFI Device Firmware:
│     Device ID:          317fd9ef3f069f67047ab1163f8c040c3853cc38
│     Current version:    0.73.8
│     Minimum Version:    0.73.8
│     Vendor:             DMI:LENOVO
│     Update State:       Success
│     Last modified:      2021-05-03 14:18
│     GUIDs:              1181d850-c3be-4b1d-985a-b28dca6ba6d8
│                         748748d3-60a0-5280-93ac-0a74d04ea947 ← UEFI\RES_{1181D850-C3BE-4B1D-985A-B28DCA6BA6D8}
│     Device Flags:       • Internal device
│                         • Updatable
│                         • System requires external power source
│                         • Supported on remote server
│                         • Needs a reboot after installation
│                         • Device is usable for the duration of the update
│   
└─UHD Graphics:
      Device ID:          5792b48846ce271fab11c4a545f7a3df0d36e00a
      Current version:    02
      Vendor:             Intel Corporation (PCI:0x8086)
      GUIDs:              b4da2a12-6b87-520c-a4be-af5b47b4c8fc ← PCI\VEN_8086&DEV_9B41&SUBSYS_17AA2292&REV_02
                          33d24107-5e36-5432-867b-627ee477f9da ← PCI\VEN_8086&DEV_9B41&SUBSYS_17AA2292
                          9779d679-0f31-5dd5-b3c2-f6ff275a9ca9 ← PCI\VEN_8086&DEV_9B41&REV_02
                          5589d313-88e5-54d3-8f4c-b2ee3882d4bc ← PCI\VEN_8086&DEV_9B41
      Device Flags:       • Internal device
                          • Cryptographic hash verification is available

```

**Additional questions**
- Operating system and version:
```
Distributor ID:	Ubuntu
Description:	Ubuntu 21.04
Release:	21.04
Codename:	hirsute
Linux 5.11.0-16-generic fwupd/fwupd#17-Ubuntu SMP Wed Apr 14 20:12:43 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

```

- Have you tried rebooting?
`Many times.`
- Is this a regression?
`I have installed many fw updates on Ubuntu 20.10 without problems. This is the first on 21.04

The update in question:
`
![Screenshot from 2021-05-01 09-55-15](https://user-images.githubusercontent.com/1281031/116890587-ca0e2e00-abfb-11eb-9018-3dac69672824.png)

**Comment by johnnyapol** [comment]
I''m hitting a very similar issue. Arch Linux, Thinkpad T490. Output of fwupdate -L is as follows:
`../efi/fwupdate.c:158:fwup_populate_update_table(): Found update fwupd-2c0665e2-fdbd-495e-b8e4-69d92b9c119a-0
../efi/fwupdate.c:158:fwup_populate_update_table(): Found update fwupd-603baf73-b997-45b5-86b4-2f981a008e18-0
../efi/fwupdate.c:564:efi_main(): No updates to process.  Called in error?
`
`~ $ pacman -Q fwupd fwupd-efi
fwupd 1.6.0-3
fwupd-efi 1.0-1
`

**Comment by crosser** [comment]
It would seem that shim got broken _again_ on hirsute. This is how my efi vars look after running fwupdmgr and reboot:
```
BootCurrent: 0002
Timeout: 0 seconds
BootOrder: 0001,0019,001A,001B,001C,001D,001E,001F,0020,0021,0022,0023,0024,0002
Boot0001* ubuntu	HD(1,GPT,6ccce482-e2c2-48ca-991e-608bee5d38af,0x800,0x100000)/File(\EFI\ubuntu\shimx64.efi)
Boot0002* Linux-Firmware-Updater	HD(1,GPT,6ccce482-e2c2-48ca-991e-608bee5d38af,0x800,0x100000)/File(\EFI\ubuntu\shimx64.efi)\.f.w.u.p.d.x.6.4...e.f.i...
Boot0010  Setup	FvFile(721c8b66-426c-4e86-8e99-3457c46ab0b9)
...
```
Note that BootCurrent points to the updater, but updater _did not run_ on boot, and I was dropped into grub right away.
I also have both "ubuntu" and "Linux firmware updater" entries in the BIOS boot menu (Enter -> F12), and again both entries drop me into grub.
Looks exactly like this problem https://github.com/rhboot/shim/issues/194 all over again.
So I, too, have this problem (hirsute on Thinkpad X1 gen 7), but I think it's not fwupdmgr's fault this time...

**Comment by superm1** [comment]
@crosser probably better not to conflate a distro problem here, especially if it's caused by shim.  I'd suggest try with SB off, and if it works, then a shim bug report in Launchpad (feel free to leave a link here though for anyone that wants to follow along)

**Comment by LinuxOnTheDesktop** [comment]
#114 is a similar seeming problem but on _Fedora_ . .

**Comment by superm1** [comment]
Which could mean it's actually upstream shim issue then too...  @vathpela, @xnox ^

**Comment by Issen007** [comment]
@crosser @superm1 I think we are on the right way here where BIOS/uEFI can't find the Update Partition after reboot.
Is it possible to verify this?

**Comment by crosser** [comment]
> @crosser probably better not to conflate a distro problem here, especially if it's caused by shim. I'd suggest try with SB off, and if it works, then a shim bug report in Launchpad (feel free to leave a link here though for anyone that wants to follow along)

@superm1 Mario, I've opened a launchpad ticket, I just thought that it could be useful to point to a potential external cause..
https://bugs.launchpad.net/ubuntu/+source/shim-signed/+bug/1929471

**Comment by crosser** [comment]
> @crosser @superm1 I think we are on the right way here where BIOS/uEFI can't find the Update Partition after reboot.
> Is it possible to verify this?

@Issen007 How can I do that? I would be happy to check and report.

**Comment by Issen007** [comment]
@crosser that was my question, can we debug the boot sequence and see why the BIOS doesn't boot correct.
I guess it doesn't find the partition or any necessary files that should be there.

**Comment by crosser** [comment]
@Issen007 as far as I understand, bios successfully loads the shim (note that both entry 1 and 2 specify `/File(\EFI\ubuntu\shimx64.efi)`, it's the shim's parameter that is different), but the shim (apparently) refuses to chain into fwupdx64.

**Comment by johnnyapol** [comment]
For what it's worth, I don't use secure boot (therefore no shim is present) and the updates still fail.

**Comment by superm1** [comment]
> For what it's worth, I don't use secure boot (therefore no shim is present) and the updates still fail.

Which is what this bug report is originally about - fwupdx64.efi launches, but the firmware fails to install it on this OEM platform.

**Comment by vathpela** [comment]
> Which could mean it's actually upstream shim issue then too... @vathpela, @xnox ^

Any chance you have a way to check if https://github.com/rhboot/shim/pull/374 fixes this?

**Comment by julian-klode** [comment]
Verification of rhboot/shim#374 failed. I'm hopeful rhboot/shim#379 will be OK, but can't know for sure yet, fixing this is out of scope for this month's Ubuntu shim update, I'm afraid.

**Comment by superm1** [comment]
I don't think it's useful to keep this bug open making this seem like it could be a firmware bug.  
It's truly a shim bug, and it will need to be fixed in shim upstream and then each of the distros need to pick it up.
It's really unfortunate that this has effectively broken all UEFI firmware updates w/ secure boot.

Ubuntu bug tracker:
https://bugs.launchpad.net/ubuntu/+source/shim/+bug/1929471

Attempts at fixing it upstream:
https://github.com/rhboot/shim/pull/379
https://github.com/rhboot/shim/pull/381

**Comment by johnnyapol** [comment]
I'm still hitting this issue without shim / secure boot. it will boot into the efi executable but then say "no updates found"/

**Comment by superm1** [comment]
Can you open your own issue with the template for debugging? There certainly is a shim bug, but you might have another problem.

**Comment by LinuxOnTheDesktop** [comment]
I still have this problem. I have had it for some nine months. Has Ubuntu (it's Ubuntu's problem, right?) actually fixed the problem? If so, the fix has yet to reach Linux Mint.

**Comment by Vernando** [comment]
I know it's a closed thread, but an open unresolved issue for me.
I found this thread here as best indication of what my problem might be, thus I add my details.

I have exactly this same described issue on 
- ThinkPad T480
- OpenSuse Leap 15.3
- Current Bios version N24ET61W (1.36)
- Already updated Bios on OpenSuse after purchase in the past
- Wanted to update to new bios version N24ET67W (1.42)
- I ran fwupdmgr refresh --force; fwupdmgr install N24ET67W.cab which did not do its job, but reported success.
- I ran the mentioned steps above of removing the two unused .cab files, removing fwupd, installing fwupd anew, repeat installation. Outcome is always the same, it reports:
```
fwupdmgr install N24ET67W.cab 
Decompressing?           [***************************************]
Authenticating?          [-                                      ]
Authenticating?          [***************************************]
Installing on System Firmware?                                   ]
Scheduling?              [***************************************]
Successfully installed firmware: The computer will be restarted automatically after updating BIOS completely . Do NOT turn  off your computer or remove the AC adaptor while update is in progress.? 

An update requires a reboot to complete. Restart now? [y|N]: 

```
- Then after the restart it does not install the bios, but moves to the regular boot sequence instead. The installed bios stays at the old unchanged version.

**Comment by devurandom** [comment]
Are you maybe experiencing https://bugzilla.redhat.com/show_bug.cgi?id=2010595 ?

**Comment by Vernando** [comment]
> Are you maybe experiencing https://bugzilla.redhat.com/show_bug.cgi?id=2010595 ?

A big thank you for this hint! 
This did the job for me. I was not aware of this at all.
With this input I could successfully update my bios to version 1.42 now.

As in the referred link suggested I needed to 
- Turn off 'Secure Boot' in the bios.
- Re-run the bios update procedure as described above.
This time it did not just report to be successful, but also did what it should do and updated the bios after the restart.
- After the successful bios-update I re-enabled 'Secure Boot' in bios again.
- Bios got updated, system started and worked well afterwards.

Really great!

My system details were:
efibootmgr        14-4.3.2
fwupd             1.5.8-150300.3.3.1
fwupdate-efi      12-11.8.2
shim              15.4-4.7.1
